smallrye-reactive-messaging-mqtt/src/main/java/io/smallrye/reactive/messaging/mqtt/session/impl/MqttClientSessionImpl.java

```diff
@@ -0,0 +1,672 @@
+import io.vertx.mqtt.MqttClient;
+import io.vertx.mqtt.messages.MqttConnAckMessage;
+import io.vertx.mqtt.messages.MqttPublishMessage;
+import io.vertx.mqtt.messages.MqttSubAckMessage;
+
+public class MqttClientSessionImpl implements MqttClientSession {
+
+    private static final Logger log = LoggerFactory.getLogger(MqttClientSessionImpl.class);
+
+    private final VertxInternal vertx;
+    private final MqttClientSessionOptions options;
+
+    // record the subscriptions
+    private final Map<String, RequestedQoS> subscriptions = new HashMap<>();
+    // record the pending subscribes
+    private final Map<Integer, LinkedHashMap<String, RequestedQoS>> pendingSubscribes = new HashMap<>();
+    // record the pending unsubscribes
+    private final Map<Integer, List<String>> pendingUnsubscribes = new HashMap<>();
+    // the provider for the reconnect delay
+    private final ReconnectDelayProvider reconnectDelay;
+
+    // the current state
+    private volatile SessionState state = SessionState.DISCONNECTED;
+    // drives to connection either to CONNECTED or DISCONNECTED
+    private volatile boolean running;
+    // subscription states
+    private final Map<String, SubscriptionState> subscriptionStates = new ConcurrentHashMap<>();
+
+    // holds the actual MQTT client connection
+    private MqttClient client;
+    // an optional reconnect timer
+    private Long reconnectTimer;
+
+    private volatile Handler<MqttPublishMessage> messageHandler;
+    private volatile Handler<Throwable> exceptionHandler;
+    private volatile Handler<SessionEvent> sessionStateHandler;
+    private volatile Handler<SubscriptionEvent> subscriptionStateHandler;
+    private volatile Handler<Integer> publishCompleteHandler;
+    private volatile Handler<Integer> publishCompletionExpirationHandler;
+    private volatile Handler<Integer> publishCompletionUnknownPacketIdHandler;
+
+    /**
+     * Create a new instance, which is not started.
+     *
+     * @param vertx The vert.x instance to use.
+     * @param options The client session options.
+     */
+    public MqttClientSessionImpl(final Vertx vertx, final MqttClientSessionOptions options) {
+        this.vertx = (VertxInternal) vertx;
+        this.options = options;
+        this.reconnectDelay = options.getReconnectDelay().createProvider();
+    }
+
+    @Override
+    public void start() {
+        this.vertx.runOnContext(x -> doStart());
+    }
+
+    @Override
+    public void stop() {
+        this.vertx.runOnContext(x -> doStop());
+    }
+
+    @Override
+    public SessionState getState() {
+        return this.state;
+    }
+
+    @Override
+    public SubscriptionState getSubscriptionState(String topicFilter) {
+        return this.subscriptionStates.get(topicFilter);
+    }
+
+    @Override
+    public MqttClientSession subscribe(Map<String, RequestedQoS> topics) {
+        final Map<String, RequestedQoS> finalTopics = new LinkedHashMap<>(topics);
```

Comment:
it would probably be better to report the success as a Future or Uni

Reply:
Same here.